components/builder-core/src/api.rs

```diff
@@ -0,0 +1,277 @@
+        Ok(g) => g,
+        Err(err) => return Err(Error::NetError(err)),
+    };
+
+    // This only makes sense if the group is complete. If the group isn't complete, return now and
+    // let the user know.
+    if group.get_state() != GroupState::Complete {
+        return Err(Error::GroupNotComplete);
+    }
+
+    let mut failed_projects = Vec::new();
+    let mut origin_map = HashMap::new();
+
+    // We can't assume that every project in the group belongs to the same origin. It's entirely
+    // possible that there are multiple origins present within the group. Because of this, there's
+    // no way to atomically commit the entire promotion at once. It's possible origin shards can be
+    // on different machines, so for now, the best we can do is partition the projects by origin,
+    // and commit each origin at once. Ultimately, it'd be nice to have a way to atomically commit
+    // the entire promotion at once, but that would require a cross-shard tool that we don't
+    // currently have.
+    for project in group.get_projects().into_iter() {
+        if project.get_state() == ProjectState::Success {
+            let ident = OriginPackageIdent::from_str(project.get_ident()).unwrap();
+
+            if channel != STABLE_CHANNEL {
+                match create_channel(ident.get_origin(), channel, session_id) {
+                    Ok(_) => (),
+                    Err(Error::NetError(err)) => {
+                        match err.get_code() {
+                            ErrCode::ENTITY_CONFLICT => (),
+                            _ => return Err(Error::NetError(err)),
+                        }
+                    }
+                    Err(e) => return Err(e),
+                }
+            }
+
+            let mut project_list = origin_map.entry(ident.get_origin().to_string()).or_insert(
+                Vec::new(),
+            );
+            project_list.push(project);
+        } else {
+            failed_projects.push(project.get_ident().to_string());
+        }
+    }
+
+    // Now that we've sorted all the projects into a HashMap keyed on origin, we can process them,
+    // one origin at a time. We do "core" first, since it's not possible to do the entire operation
+    // atomically. Instead, we prioritize the core origin since it's the most important one.
+    if let Some(core_projects) = origin_map.remove("core") {
+        let promote_result = do_group_promotion(channel, core_projects, "core", session_id);
+        if promote_result.is_err() {
+            return promote_result;
+        }
+    }
+
+    for (origin, projects) in origin_map.iter() {
+        let promote_result = do_group_promotion(channel, projects.to_vec(), &origin, session_id);
+        if promote_result.is_err() {
+            return promote_result;
+        }
+    }
+
+    if failed_projects.is_empty() {
+        Ok(())
+    } else {
+        let ujgp = UnsuccessfulJobGroupPromote {
```

Comment:
When I wrote this, I called it "unsuccessful" because the entire job group wasn't promoted successfully.  I can rename it to reflect the partial failure/partial success.  I changed the error message that `hab` returns to address your other questions.